sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/datastore/DataStoreV1Table.java

```diff
@@ -0,0 +1,479 @@
+    JSONObject properties = table.getProperties();
+    if (properties.containsKey(KEY_FIELD_PROPERTY)) {
+      String field = properties.getString(KEY_FIELD_PROPERTY);
+      checkArgument(
+          field != null && !field.isEmpty(), "'%s' property cannot be null.", KEY_FIELD_PROPERTY);
+      keyField = field;
+    } else {
+      keyField = DEFAULT_KEY_FIELD;
+    }
+    // TODO: allow users to specify a namespace in a location string.
+    String location = table.getLocation();
+    assert location != null;
+    Matcher matcher = locationPattern.matcher(location);
+    checkArgument(
+        matcher.matches(),
+        "DataStoreV1 location must be in the following format: 'projectId/kind'");
+
+    this.projectId = matcher.group("projectId");
+    this.kind = matcher.group("kind");
+  }
+
+  @Override
+  public PCollection<Row> buildIOReader(PBegin begin) {
+    Query.Builder q = Query.newBuilder();
+    q.addKindBuilder().setName(kind);
+    Query query = q.build();
+
+    DatastoreV1.Read readInstance =
+        DatastoreIO.v1().read().withProjectId(projectId).withQuery(query);
+
+    PCollection<Entity> readEntities = readInstance.expand(begin);
+
+    return readEntities.apply(EntityToRow.create(getSchema(), keyField));
+  }
+
+  @Override
+  public POutput buildIOWriter(PCollection<Row> input) {
+    return input
+        .apply(RowToEntity.create(keyField, kind))
+        .apply(DatastoreIO.v1().write().withProjectId(projectId));
+  }
+
+  @Override
+  public IsBounded isBounded() {
+    return IsBounded.BOUNDED;
+  }
+
+  @Override
+  public BeamTableStatistics getTableStatistics(PipelineOptions options) {
+    long count =
+        DatastoreIO.v1().read().withProjectId(projectId).getNumEntities(options, kind, null);
+
+    if (count < 0) {
+      return BeamTableStatistics.BOUNDED_UNKNOWN;
+    }
+
+    return BeamTableStatistics.createBoundedTableStatistics((double) count);
+  }
+
+  /**
+   * A {@code PTransform} to perform a conversion of {@code PCollection<Entity>} to {@code
+   * PCollection<Row>}.
+   */
+  public static class EntityToRow extends PTransform<PCollection<Entity>, PCollection<Row>> {
+    private final Schema schema;
+    private final String keyField;
+
+    private EntityToRow(Schema schema, String keyField) {
+      this.schema = schema;
+      this.keyField = keyField;
+
+      if (schema.getFieldNames().contains(keyField)
+          && !schema.getField(keyField).getType().getTypeName().equals(TypeName.BYTES)) {
+        throw new IllegalStateException(
+            "Field `"
+                + keyField
+                + "` should of type `VARBINARY`. Please change the type or specify a field to store the KEY value in via TableProperties.");
+      }
+    }
+
+    /**
+     * Create a PTransform instance.
+     *
+     * @param schema {@code Schema} of the target row.
+     * @return {@code PTransform} instance for Entity to Row conversion.
+     */
+    public static EntityToRow create(Schema schema) {
```

Comment:
Looks like nothing but tests call this method. Can you simplify by just dropping?

Reply:
Done.